KAFKA-5526: Additional `--describe` views for ConsumerGroupCommand (KIP-175)

vahidhashemian opened the pull request:
The `--describe` option of ConsumerGroupCommand is expanded, as proposed in [KIP-175](https://cwiki.apache.org/confluence/display/KAFKA/KIP-175%3A+Additional+%27--describe%27+views+for+ConsumerGroupCommand), to support:
* `--describe` or `--describe --offsets`: listing of current group offsets
* `--describe --members` or `--describe --members --verbose`: listing of group members
* `--describe --state`: group status

Example: With a single partition topic `test1` and a double partition topic `test2`, consumers `consumer1` and `consumer11` subscribed to `test`, consumers `consumer2` and `consumer22` and `consumer222` subscribed to `test2`, and all consumers belonging to group `test-group`, this is an output example of the new options above for `test-group`:

```
--describe, or --describe --offsets:

TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                     HOST            CLIENT-ID
test2           0          0               0               0               consumer2-bad9496d-0889-47ab-98ff-af17d9460382  /127.0.0.1      consumer2
test2           1          0               0               0               consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411 /127.0.0.1      consumer22
test1           0          0               0               0               consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf  /127.0.0.1      consumer1
```

```
--describe --members

CONSUMER-ID                                      HOST            CLIENT-ID       #PARTITIONS
consumer2-bad9496d-0889-47ab-98ff-af17d9460382   /127.0.0.1      consumer2       1
consumer222-ed2108cd-d368-41f1-8514-5b72aa835bcc /127.0.0.1      consumer222     0
consumer11-dc8295d7-8f3f-4438-9b11-7270bab46760  /127.0.0.1      consumer11      0
consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411  /127.0.0.1      consumer22      1
consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf   /127.0.0.1      consumer1       1
```

```
--describe --members --verbose

CONSUMER-ID                                      HOST            CLIENT-ID       #PARTITIONS     ASSIGNMENT
consumer2-bad9496d-0889-47ab-98ff-af17d9460382   /127.0.0.1      consumer2       1               test2(0)
consumer222-ed2108cd-d368-41f1-8514-5b72aa835bcc /127.0.0.1      consumer222     0               -
consumer11-dc8295d7-8f3f-4438-9b11-7270bab46760  /127.0.0.1      consumer11      0               -
consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411  /127.0.0.1      consumer22      1               test2(1)
consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf   /127.0.0.1      consumer1       1               test1(0)
```

```
--describe --state

COORDINATOR (ID)         ASSIGNMENT-STRATEGY       STATE                #MEMBERS
localhost:9092 (0)       range                     Stable               5
```

Note that this PR also addresses the issue reported in [KAFKA-6158](https://issues.apache.org/jira/browse/KAFKA-6158) by dynamically setting the width of columns `TOPIC`, `CONSUMER-ID`, `HOST`, `CLIENT-ID` and `COORDINATOR (ID)`. This avoid truncation of column values when they go over the current fixed width of these columns.

The code has been restructured to better support testing of individual values and also the console output. Unit tests have been updated and extended to take advantage of this restructuring.

vahidhashemian committed:
KAFKA-5526: Additional `--describe` views for ConsumerGroupCommand (KIP-175)

The `--describe` option of ConsumerGroupCommand is expanded to support:
* `--describe` or `--describe --offsets`: listing of current group offsets
* `--describe --members` or `--describe --members --verbose`: listing of group members
* `--describe --state`: group status

Example: With a single partition topic `test1` and a double partition topic `test2`, consumers `consumer1` and `consumer11` subscribed to `test`, consumers `consumer2` and `consumer22` and `consumer222` subscribed to `test2`, and all consumers belonging to group `test-group`, this is an output example of the new options above for `test-group`:

```
--describe, or --describe --offsets:

TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                     HOST            CLIENT-ID
test2           0          0               0               0               consumer2-bad9496d-0889-47ab-98ff-af17d9460382  /127.0.0.1      consumer2
test2           1          0               0               0               consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411 /127.0.0.1      consumer22
test1           0          0               0               0               consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf  /127.0.0.1      consumer1
```

```
--describe --members

CONSUMER-ID                                      HOST            CLIENT-ID       #PARTITIONS
consumer2-bad9496d-0889-47ab-98ff-af17d9460382   /127.0.0.1      consumer2       1
consumer222-ed2108cd-d368-41f1-8514-5b72aa835bcc /127.0.0.1      consumer222     0
consumer11-dc8295d7-8f3f-4438-9b11-7270bab46760  /127.0.0.1      consumer11      0
consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411  /127.0.0.1      consumer22      1
consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf   /127.0.0.1      consumer1       1
```

```
--describe --members --verbose

CONSUMER-ID                                      HOST            CLIENT-ID       #PARTITIONS     ASSIGNMENT
consumer2-bad9496d-0889-47ab-98ff-af17d9460382   /127.0.0.1      consumer2       1               test2(0)
consumer222-ed2108cd-d368-41f1-8514-5b72aa835bcc /127.0.0.1      consumer222     0               -
consumer11-dc8295d7-8f3f-4438-9b11-7270bab46760  /127.0.0.1      consumer11      0               -
consumer22-c45e6ee2-0c7d-44a3-94a8-9627f63fb411  /127.0.0.1      consumer22      1               test2(1)
consumer1-d51b0345-3194-4305-80db-81a68fa6c5bf   /127.0.0.1      consumer1       1               test1(0)
```

```
--describe --state

ASSIGNMENT-STRATEGY       STATE                #MEMBERS
range                     Stable               5
```

Note that this PR also addresses the issue reported in [KAFKA-6158](https://issues.apache.org/jira/browse/KAFKA-6158) by dynamically setting the width of columns `TOPIC`, `CONSUMER-ID`, `HOST` and `CLIENT-ID`. This avoid truncation of column values when they go over the current fixed width of these columns.

vahidhashemian commented:
@hachikuji, please take a look when you get a chance. Thanks. Test failures seem to be unrelated.

omkreddy commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -464,6 +558,28 @@ object ConsumerGroupCommand extends Logging {
+              MemberAssignmentState(group, Some(consumer.consumerId), Some(consumer.host), Some(consumer.clientId),
+                                    Some(consumer.assignment.length),
+                                    if (verbose) Some(consumer.assignment) else None)
+            }.toList)
```
redundant toList

omkreddy reviewed (commented):
LGTM.

vahidhashemian committed:
Minor edits

vahidhashemian commented:
@omkreddy, thanks for the review. I addressed your comment in the new commit.

brettrann commented:
With that assignment column, how is it going to look when, for example, there's a single consumer assigned 100s of partitions?

vahidhashemian commented:
@brettrann, it would print all assigned partitions in the format `topic(partitions), ...`. Note that the `ASSIGNMENT` column is the last column (not bounded by a fixed width) and is included only in the `--verbose` option. Not sure if there is a better way of reporting all assigned partitions; if you can think of any please let me know. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -53,53 +55,27 @@ object ConsumerGroupCommand extends Logging {
     val consumerGroupService = {
       if (opts.useOldConsumer) {
-        System.err.println("Note: This will only show information about consumers that use ZooKeeper (not those using the Java consumer API).\n")
+        Console.err.println("Note: This will only show information about consumers that use ZooKeeper (not those using the Java consumer API).\n")
```
From the old code, but are the newlines needed since we use `println`?

vahidhashemian replied:
It can be removed. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
                                                 consumerId: Option[String], host: Option[String],
                                                 clientId: Option[String], logEndOffset: Option[Long])
 
+  protected case class MemberAssignmentState(group: String, consumerId: Option[String], host: Option[String],
```
Having so many `Option` fields seems less than ideal unless every permutation actually needs to be represented. Maybe it could be simplified by encapsulating the optional state into a single case class? For example:
```scala
case class AssignmentState(consumerId: String, host: String, clientId: String, numPartitions: Int, assignment: List[TopicPartition])
case class MemberAssignmentState(group: String, assignment: Option[AssignmentState])
```
Same for `GroupState` below.

vahidhashemian replied:
Thanks for the suggestion. I think for almost all fields the `Option` definition does not make sense (based on how we expect the output for `--members` and `--state` cases). I'll update and simplify this.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
 
-    def describeGroup(): (Option[String], Option[Seq[PartitionAssignmentState]]) = {
-      collectGroupAssignment(opts.options.valueOf(opts.groupOpt))
+    private def checkGroup(group: String, state: Option[String], numRows: Option[Int]): Boolean = {
```
Could we choose a more descriptive name? Maybe `isGroupActive` or something like that? Also, not really obvious what purpose `numRows` is serving since it's use is a bit indirect.

vahidhashemian replied:
I'll try to address these in the next commit.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
-      collectGroupAssignment(opts.options.valueOf(opts.groupOpt))
+    private def checkGroup(group: String, state: Option[String], numRows: Option[Int]): Boolean = {
+      var printList = true
+      numRows.isDefined match {
```
Wouldn't it be more natural to match on `numRows` directly?

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
+    }
+
+    private def size(colOpt: Option[Seq[Object]]): Option[Int] =
+      colOpt.flatMap(col => Try(col.size).toOption)
```
Isn't this just `colOpt.map(_.size)`?

vahidhashemian replied:
Yes, thanks for simplifying it :)

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
+            maxHostLen = Math.max(maxHostLen, consumerAssignment.host.getOrElse(MISSING_COLUMN_VALUE).length)
+        }
+
+        print(s"\n%${-maxTopicLen}s %-10s %-15s %-15s %-15s %${-maxConsumerIdLen}s "
```
Not sure if there's a nicer way to do this. I know it's also possible to include the formatter behind the substitution. For example: `$foo%-15s` would apply the formatter to whatever is substituted by `foo`.

vahidhashemian replied:
Would that help with dynamically set column widths (e.g. `maxTopicLen` or `maxConsumerIdLen`)? It seems applying it would further complicate the formatter string. Please advise if I misunderstood

hachikuji replied:
I was referring to the mix of string interpolation and format, which is a bit odd. It seems like interpolation already provides the formatting facilities you're using here.

vahidhashemian replied:
I'm still not sure if I understood your point. Since the column widths is now going to be dynamically set for a few columns, I looked up how it can be done via interpolation and results like [this](https://stackoverflow.com/questions/19339935/printing-with-variable-length-width) came up, hence `%${-maxTopicLen}s`, etc. It's the formatter itself that's dynamic (to address [KAFKA-6158](https://issues.apache.org/jira/browse/KAFKA-6158)).

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,141 @@ object ConsumerGroupCommand extends Logging {
+          print(s"%${-maxHostLen}s %s".format("HOST", "CLIENT-ID"))
+        println()
+
+        assignments.get.foreach { consumerAssignment =>
```
Seems like the `get` is safe here because of `checkGroup`, but the logic is a little indirect. I wonder if it would be simpler to have a direct `match` on `assignments`.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -449,37 +543,55 @@ object ConsumerGroupCommand extends Logging {
-                }
+            val rowsWithoutConsumer = offsets.filterKeys(!assignedTopicPartitions.contains(_)).flatMap {
+              case (topicPartition, offset) =>
+                val topicAndPartition = new TopicAndPartition(topicPartition)
```
Not a big deal, but maybe we can just use `TopicPartition` and do the conversion only in the zk path.

vahidhashemian replied:
Would it be okay to leave this improvement out? If I understood your suggestion correctly (use `TopicPartition` everywhere except when working with ZK), it impacts quite a few places in the code, and can be done in separate PR as it's not directly related to this one. If you think it should be done as part of this PR, I'll do that in a separate commit. Thanks.

hachikuji replied:
Yes, that's fine. Thanks.

vahidhashemian replied:
Thanks. I'll make sure I'll do this change in a separate PR.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -449,37 +543,55 @@ object ConsumerGroupCommand extends Logging {
+      val consumerGroupSummary = adminClient.describeConsumerGroup(group, opts.options.valueOf(opts.timeoutMsOpt))
+      (Some(consumerGroupSummary.state),
+        consumerGroupSummary.consumers match {
+          case None =>
```
If `None` maps to `None`, that's a good sign that we can use a `map` instead.

hachikuji commented on core/src/test/scala/unit/kafka/admin/DescribeConsumerGroupTest.scala:
```diff
@@ -136,138 +141,537 @@ class DescribeConsumerGroupTest extends KafkaServerTestHarness {
+
+    for (describeType <- describeTypes) {
+      // run one consumer in the group consuming from a single-partition topic
+      addConsumerGroupExecutor(new ConsumerGroupExecutor(brokerList, 1, group, topic))
```
I guess you're following a pre-existing pattern, but running a separate group seems like overkill for this test case.

vahidhashemian replied:
You're right. This can be moved out of the loop. Thanks.

hachikuji replied:
I was more wondering if it needed to be in the test at all since we don't actually use it.

vahidhashemian replied:
You're absolutely right. No need to run any group here. Thanks.

hachikuji commented on core/src/test/scala/unit/kafka/admin/DescribeConsumerGroupTest.scala:
```diff
@@ -136,138 +141,537 @@ class DescribeConsumerGroupTest extends KafkaServerTestHarness {
+    }, s"Expected a 'Stable' group status, rows and valid member information for group $group.")
+
+    val (state, assignments) = service.collectGroupMembers(true)
+    assertTrue(s"Expected a topic partition assigned to the single group member for group $group",
```
You could also do a `match` and `fail` the None case. Similar below.

vahidhashemian replied:
I updated this one instance. Do you suggest changing the `waitUntilTrue`constructs to this format too?

hachikuji replied:
Yes, that makes sense.

vahidhashemian replied:
For now, I modified the `waitUntilTrue` construct in the same unit test. If that sounds good to you (it seems less readable to me) I'll update the rest similarly. Thanks.

hachikuji commented on core/src/test/scala/unit/kafka/admin/DescribeConsumerGroupTest.scala:
```diff
@@ -136,138 +141,537 @@ class DescribeConsumerGroupTest extends KafkaServerTestHarness {
+        state.isDefined &&
+        state.get.state == Some("Stable") &&
+        state.get.numMembers == Some(1) &&
+        state.get.assignmentStrategy == Some("roundrobin")
```
Curiously, this is the only assertion we have in this test case on the assignment strategy. Is it significant that it uses round-robin or could it just as well be range?

vahidhashemian replied:
We're just testing here that changing the strategy from default reflects in the group state description result.

hachikuji replied:
Fair enough. Maybe a comment or perhaps this should even be a separate test case.

vahidhashemian replied:
Sure, I'll create a separate unit test for this.

hachikuji commented on core/src/test/scala/unit/kafka/admin/DescribeConsumerGroupTest.scala:
```diff
@@ -281,15 +685,27 @@ class DescribeConsumerGroupTest extends KafkaServerTestHarness {
     oldConsumers += new OldConsumer(Whitelist(topic), consumerProps)
   }
+
+  def addConsumerGroupService(service: ConsumerGroupService): ConsumerGroupService = {
```
nit: you could probably just construct the service in this method if you pass the command options.

hachikuji commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1503,6 +1503,36 @@ object TestUtils extends Logging {
+  def grabConsoleError(f: => Unit) : String = {
+    val err = new ByteArrayOutputStream
+    try scala.Console.withErr(err)(f)
+    finally scala.Console.err.flush
```
nit: `flush` mutates, so should have parenthesis.

hachikuji commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1503,6 +1503,36 @@ object TestUtils extends Logging {
+    try {
+      scala.Console.withOut(out) {
+        scala.Console.withErr(err) {
+          f
```
nit: This would look a little less odd if we made it a normal function and invoked with parenthesis.

hachikuji reviewed (commented):
Thanks for the patch. I appreciate the thorough test cases. Left a few (mostly stylistic) comments.

vahidhashemian committed:
Minor updates based on the feedback received

vahidhashemian reviewed (commented):
@hachikuji Thanks for the thorough review. I tried to address most of your comments in the new commit. For a couple others your clarification is much appreciated. Thanks.

brettrann commented:
@vahidhashemian i don't have a better suggestion no, just wanted to check. This is why we have less and -S, no probs.  I'm looking forward to these changes!

vahidhashemian committed:
additional minor updates based on the feedback

vahidhashemian reviewed (commented):
@hachikuji thanks for another review. Please take a look a look at the new commit, and also my responses. Still need some clarification on that interpolation comment. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,154 @@ object ConsumerGroupCommand extends Logging {
+    private def dataExistsForGroup(group: String, state: Option[String], numRows: Option[Int]): Boolean = {
+      // numRows contains the number of data rows, if any, compiled from the API call in the caller method.
+      // if it's undefined or 0, there is no group metadata to display.
+      var printList = true
```
nit: it would be a little more idiomatic to return directly from the `match`.

vahidhashemian replied:
Sure I'll try that in the next commit. Hopefully readability would not suffer.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,154 @@ object ConsumerGroupCommand extends Logging {
 
-    def describeGroup(): (Option[String], Option[Seq[PartitionAssignmentState]]) = {
-      collectGroupAssignment(opts.options.valueOf(opts.groupOpt))
+    private def dataExistsForGroup(group: String, state: Option[String], numRows: Option[Int]): Boolean = {
```
Maybe a better name is `shouldPrintMemberState` or something like that? The third argument still feels a bit odd (in particular when we have to pass `Some(1)` in `printState` below).

vahidhashemian replied:
I'll rename the method. One challenge is mixing the logic for old and new consumer groups. This is why I used the third argument as is instead of a boolean to keep the logic simple (e.g. simply using a boolean like `rowsExist` would add to the complexity). That said, if you think it's necessary I'll work on refactoring this method to see if I can reduce the oddity.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -807,29 +937,49 @@ object ConsumerGroupCommand extends Logging {
+          CommandLineUtils.printUsageAndDie(parser,
+              s"Command may include exactly one $describeOpt sub-action: $membersOpt, $offsetsOpt, $stateOpt.")
+      }
+      if (options.has(verboseOpt) && !membersOptPresent)
```
Hmm.. This check seems like overkill. Verbose could just be a no-op if it's provided for an argument which doesn't have a verbose view. No strong opinion though.

vahidhashemian replied:
Sure, it wasn't covered like this in the KIP. But hopefully that's not a big deal.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -807,29 +937,49 @@ object ConsumerGroupCommand extends Logging {
+        if (!describeOptPresent)
+          CommandLineUtils.printUsageAndDie(parser,
+              s"Options $membersOpt, $offsetsOpt, and $stateOpt may only be used with $describeOpt.")
+        if (subActions > 1)
```
The options parser has a `mutuallyExclusive` function that might already provide this.

vahidhashemian replied:
Good suggestion, though I think the problem is the call to `mutuallyExclusive` has to be made before `parser.parse()` (which is well before this point of code, where we've already used some logic around old vs. new consumer, ...). Simply using 
```
parser.mutuallyExclusive(membersOpt, offsetsOpt, stateOpt)
```
instead of this `if` block does not seem to work. Am I missing something?

hachikuji replied:
Yeah, I assumed that if you added the `mutuallyExclusive` clause, then parsing itself would fail and we could leave this check out.

vahidhashemian replied:
Let me move it up before `parse`. It would still report the same mutually exclusive error for when old consumer is used. But perhaps it's okay. Let me know if you think otherwise. Thanks.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -165,12 +124,154 @@ object ConsumerGroupCommand extends Logging {
                                                 clientId: Option[String], logEndOffset: Option[Long])
 
+  protected case class MemberAssignmentState(group: String, consumerId: String, host: String, clientId: String,
+                                             numPartitions: Int, assignment: Option[List[TopicPartition]])
```
Seems like you could just also include the assignment and leave it up to local logic whether to print it or not.

vahidhashemian replied:
Yes, I think so. I used `Option` is case we we need to distinguish between the non-`verbose` option and the `verbose` option where there is no assignment to a member. I think for now there is no use case for that. I'll update it.

hachikuji reviewed (commented):
Thanks for the updates. A few more minor comments.

vahidhashemian committed:
minor updates

vahidhashemian reviewed (commented):
@hachikuji Thanks for another look at the PR. I tried to address your comments inline, and in the new commit.

vahidhashemian committed:
Made use of `OptionParser.mutuallyExclusive(...)`

Also modified the default behavior of newly introduced methods in `ConsumerGroupService`

vahidhashemian commented:
The failed test is unrelated.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -449,37 +550,50 @@ object ConsumerGroupCommand extends Logging {
+      )
+    }
+
+    override def collectGroupState(): Option[GroupState] = {
```
When debugging a problem, it is often useful to know who the coordinator of a group is. Would it be reasonable to add that to the state output? I don't think we'd need a revote for a minor change like that.

vahidhashemian replied:
Sure, I was going to add it as part of the code change for [KIP-181](https://cwiki.apache.org/confluence/display/KAFKA/KIP-181%3A+DescribeGroups+Protocol+Enhancement) (still in draft mode), but I guess I can update KIP-175 with this minor change and submit a new commit and add coordinator id to to `ConsumerGroupSummary` class.

hachikuji replied:
Yeah, I'd suggest just doing it here.

vahidhashemian replied:
Sure, just noticed `coordinator` is already in `ConsumerGroupSummary`. So not much will need to change to get it done.

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -795,6 +915,12 @@ object ConsumerGroupCommand extends Logging {
                              .withRequiredArg()
                              .describedAs("number-of-offsets")
                              .ofType(classOf[Long])
+    val membersOpt = parser.accepts("members", MembersDoc)
```
I did some messing around with joptsimple arguments. There is an `availableIf` function which could be used in this case to declare the dependence on the `--describe` option. Unfortunately, the library is not clever enough to show this dependence in the help message, so all arguments are still organized in a flat list. Since this command is getting complex, a better help message would be a good improvement, but I guess we can take that up in a separate jira.

hachikuji replied:
Maybe if we just had a way to show a few common examples, it would be a big help.

vahidhashemian replied:
Sounds good, I'll try to make use of `availableIf` and also provide examples in the help message.

vahidhashemian committed:
Added coordinator id to describe state output.
Also, made better use joptsimple functionality wrt. enfocring command options.

vahidhashemian commented:
@hachikuji, I tried to address your recent suggestions in the new commit. Please take a look when you have a chance. The failed unit test should be unrelated, and could be another instance of [KAFKA-6118](https://issues.apache.org/jira/browse/KAFKA-6118). Thanks!

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -245,8 +245,8 @@ object ConsumerGroupCommand extends Logging {
-        print("\n%-25s %-20s %s".format("ASSIGNMENT-STRATEGY", "STATE", "#MEMBERS"))
-        print("\n%-25s %-20s %s".format(state.get.assignmentStrategy, state.get.state, state.get.numMembers))
+        print("\n%-20s %-25s %-20s %s".format("COORDINATOR", "ASSIGNMENT-STRATEGY", "STATE", "#MEMBERS"))
+        print("\n%-20s %-25s %-20s %s".format(state.get.coordinator.id, state.get.assignmentStrategy, state.get.state, state.get.numMembers))
```
Thanks for adding this. Do you think we should also print host and port? For example, maybe something like `host:port (id)`

vahidhashemian replied:
Sure, do you think the output of `Node.toString()` works? Or its too verbose?

hachikuji replied:
I was thinking about that. The addition of `rack` is kind of annoying since it will usually be null and it seems not too useful in this context. Maybe we could do `{host}:{port} (id:{id})`?

vahidhashemian replied:
Yeah, if that's okay I'll use `COORDINATOR (ID)` as the column header so we don't need to repeat `id:` in each column cell. I'll probably create a local function to get the output `{host}:{port} ({id})` since now we need to make sure the content does not overflow the width, and extend the width dynamically if necessary.

vahidhashemian replied:
Sorry, ignore that. There is only one row. So we should be okay :)

vahidhashemian committed:
Improved the coordinator value reported by describe group state

hachikuji commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -245,8 +245,10 @@ object ConsumerGroupCommand extends Logging {
       if (shouldPrintMemberState(group, Some(state.get.state), Some(1))) {
-        print("\n%-20s %-25s %-20s %s".format("COORDINATOR", "ASSIGNMENT-STRATEGY", "STATE", "#MEMBERS"))
-        print("\n%-20s %-25s %-20s %s".format(state.get.coordinator.id, state.get.assignmentStrategy, state.get.state, state.get.numMembers))
+        val coordinator = state.get.coordinator.host + ":" + state.get.coordinator.port + " (" + state.get.coordinator.idString + ")"
```
nit: since we're in scala, maybe we can use string interpolation. Also, doesn't the `state.get` seem unsafe? Are we missing a `match`?

vahidhashemian replied:
Unlike `collectGroupOffsets` and `collectGroupMembers`, `collectGroupState` always returns content (no `NONE` value). I guess it makes more sense to change the output from `Option[GroupState]` to `GroupState`. What do you think?

hachikuji replied:
Yeah, if the `None` case is not actually possible, then we should not use `Option`.

vahidhashemian replied:
The new commit has the change I mentioned above.

vahidhashemian committed:
Minor fix and improvements

hachikuji commented:
Unfortunately the examples don't render too nicely. Here's how I see them on my giant display:
```
--offsets                               Describe the group and list all topic
                                          partitions in the group along with
                                          their offset lag. This is the
                                          default sub-action of and may be
                                          used with '--describe' and '--
                                          bootstrap-server' options only.
                                        Example: --bootstrap-server localhost:
                                          9092 --describe --group group1 --
                                          offsets
```
I guess it's fine for now and we can open a jira to improve it. Some other tools use argparse4j. Seems like it has some nicer options.

hachikuji reviewed (approved):
LGTM. Thanks for the KIP and the patch iterations. I'll merge tomorrow as long as there are no problems with the testing.

Syed-SnapLogic commented:
What release of Kafka contains this fix?

hachikuji commented:
@Syed-SnapLogic It will go into 1.1.0, which is due early next year.

Syed-SnapLogic commented:
Thanks
